File: ru.kmorozov.gbd.desktop/src/main/java/ru/kmorozov/gbd/core/config/CommandLineOptions.java
package ru.kmorozov.gbd.core.config;

import org.apache.commons.cli.*;
import ru.kmorozov.gbd.core.config.options.CtxOptions;
import ru.kmorozov.gbd.core.loader.LocalFSStorage;

/**
 * Created by km on 06.12.2015.
 */
public class CommandLineOptions implements IGBDOptions {

    private static final String OPTION_BOOKID_SHORT = "i";
    private static final String OPTION_BOOKID_LONG = "bookId";
    private static final String OPTION_OUTDIR_SHORT = "o";
    private static final String OPTION_OUTDIR_LONG = "out";
    private static final String OPTION_PROXY_FILE_SHORT = "p";
    private static final String OPTION_PROXY_FILE_LONG = "proxy";
    private static final String OPTION_WIDTH_SHORT = "w";
    private static final String OPTION_WIDTH_LONG = "width";
    private static final String OPTION_IMG_RELOAD_SHORT = "r";
    private static final String OPTION_IMG_RELOAD_LONG = "reload";
    private static final String OPTION_SECURE_MODE_SHORT = "s";
    private static final String OPTION_SECURE_MODE_LONG = "secure";
    private static final String OPTION_PDF_MODE_SHORT = "x";
    private static final String OPTION_PDF_MODE_LONG = "pdf";
    private static final String OPTION_CTX_MODE_SHORT = "c";
    private static final String OPTION_CTX_MODE_LONG = "ctx";

    private CommandLine commandLine;

    public CommandLineOptions(String[] commandLineArguments) {
        CommandLineParser cmdLineParser = new DefaultParser();
        Options options = new Options();

        Option option = new Option(CommandLineOptions.OPTION_BOOKID_SHORT, CommandLineOptions.OPTION_BOOKID_LONG, true, "Book Id");
        option.setArgs(1);
        option.setOptionalArg(false);
        option.setArgName("BookId ");
        options.addOption(option);

        option = new Option(CommandLineOptions.OPTION_OUTDIR_SHORT, CommandLineOptions.OPTION_OUTDIR_LONG, true, "output dir");
        option.setArgs(1);
        option.setOptionalArg(false);
        option.setArgName("output directory ");
        options.addOption(option);

        option = new Option(CommandLineOptions.OPTION_PROXY_FILE_SHORT, CommandLineOptions.OPTION_PROXY_FILE_LONG, true, "Proxy list file");
        option.setArgs(1);
        option.setOptionalArg(false);
        option.setArgName("proxy options ");
        options.addOption(option);

        option = new Option(CommandLineOptions.OPTION_WIDTH_SHORT, CommandLineOptions.OPTION_WIDTH_LONG, true, "Width");
        option.setArgs(1);
        option.setOptionalArg(true);
        option.setArgName("Width ");
        options.addOption(option);

        option = new Option(CommandLineOptions.OPTION_IMG_RELOAD_SHORT, CommandLineOptions.OPTION_IMG_RELOAD_LONG, true, "Reload images");
        option.setArgs(0);
        option.setOptionalArg(true);
        option.setArgName("Reload images ");
        options.addOption(option);

        option = new Option(CommandLineOptions.OPTION_SECURE_MODE_SHORT, CommandLineOptions.OPTION_SECURE_MODE_LONG, true, "Secure mode");
        option.setArgs(0);
        option.setOptionalArg(true);
        option.setArgName("Secure mode ");
        options.addOption(option);

        option = new Option(CommandLineOptions.OPTION_PDF_MODE_SHORT, CommandLineOptions.OPTION_PDF_MODE_LONG, true, "Pdf mode");
        option.setArgs(1);
        option.setOptionalArg(true);
        option.setArgName("Pdf mode ");
        options.addOption(option);

        option = new Option(CommandLineOptions.OPTION_CTX_MODE_SHORT, CommandLineOptions.OPTION_CTX_MODE_LONG, true, "CTX mode");
        option.setArgs(2);
        option.setOptionalArg(true);
        option.setArgName("CTX mode ");
        options.addOption(option);

        try {
            this.commandLine = cmdLineParser.parse(options, commandLineArguments);
        } catch (ParseException e) {
            e.printStackTrace();
        }
    }

    private String getStringOptionValue(String optionName) {
        return this.commandLine.hasOption(optionName) && null != this.commandLine.getOptionValues(optionName) && 1 == this.commandLine.getOptionValues(optionName).length
                ? this.commandLine.getOptionValues(optionName)[0] : null;
    }

    private String[] getStringOptionValues(String optionName) {
        return this.commandLine.hasOption(optionName) && null != this.commandLine.getOptionValues(optionName)
                ? this.commandLine.getOptionValues(optionName) : null;
    }

    private int getIntOptionValue(String optionName) {
        return this.commandLine.hasOption(optionName) && 1 == this.commandLine.getOptionValues(optionName).length ? Integer.parseInt(this.commandLine.getOptionValues(optionName)[0]) : 0;
    }

    private boolean getBoolOptionValue(String optionName) {
        return this.commandLine.hasOption(optionName);
    }

    @Override
    public String getBookId() {
        return this.getStringOptionValue(CommandLineOptions.OPTION_BOOKID_SHORT);
    }

    @Override
    public IStorage getStorage() {
        return new LocalFSStorage(this.getStringOptionValue(CommandLineOptions.OPTION_OUTDIR_SHORT));
    }

    @Override
    public String getProxyListFile() {
        return this.getStringOptionValue(CommandLineOptions.OPTION_PROXY_FILE_SHORT);
    }

    @Override
    public int getImageWidth() {
        return this.getIntOptionValue(CommandLineOptions.OPTION_WIDTH_SHORT);
    }

    @Override
    public boolean reloadImages() {
        return this.getBoolOptionValue(CommandLineOptions.OPTION_IMG_RELOAD_SHORT);
    }

    @Override
    public boolean secureMode() {
        return this.getBoolOptionValue(CommandLineOptions.OPTION_SECURE_MODE_SHORT);
    }

    @Override
    public String pdfOptions() {
        return this.getStringOptionValue(CommandLineOptions.OPTION_PDF_MODE_SHORT);
    }

    @Override
    public CtxOptions ctxOptions() {
        final String[] ctxOpts = this.getStringOptionValues(CommandLineOptions.OPTION_CTX_MODE_SHORT);
        return ctxOpts == null || ctxOpts.length != 2 ? CtxOptions.DEFAULT_CTX_OPTIONS : new CtxOptions(ctxOpts[0], ctxOpts[1]);
    }
}

File: ru.kmorozov.gbd.desktop/src/main/java/ru/kmorozov/gbd/core/config/CommandLineOptions.kt
package ru.kmorozov.gbd.core.config

import org.apache.commons.cli.CommandLine
import org.apache.commons.cli.DefaultParser
import org.apache.commons.cli.Option
import org.apache.commons.cli.Options
import ru.kmorozov.gbd.core.config.options.AuthOptions
import ru.kmorozov.gbd.core.config.options.CtxOptions
import ru.kmorozov.gbd.core.loader.LocalFSStorage

/**
 * Created by km on 06.12.2015.
 */
class CommandLineOptions(commandLineArguments: Array<String>) : IGBDOptions {

    private val commandLine: CommandLine

    override val bookId: String
        get() = getStringOptionValue(OPTION_BOOKID_SHORT)

    override val storage: IStorage
        get() = LocalFSStorage.getStorage(getStringOptionValue(OPTION_OUTDIR_SHORT))

    override val proxyListFile: String
        get() = getStringOptionValue(OPTION_PROXY_FILE_SHORT)

    override val imageWidth: Int
        get() = getIntOptionValue(OPTION_WIDTH_SHORT)

    init {
        val cmdLineParser = DefaultParser()
        val options = Options()

        var option = Option(OPTION_BOOKID_SHORT, OPTION_BOOKID_LONG, true, "Book Id")
        option.args = 1
        option.setOptionalArg(false)
        option.argName = "BookId "
        options.addOption(option)

        option = Option(OPTION_OUTDIR_SHORT, OPTION_OUTDIR_LONG, true, "output dir")
        option.args = 1
        option.setOptionalArg(false)
        option.argName = "output directory "
        options.addOption(option)

        option = Option(OPTION_PROXY_FILE_SHORT, OPTION_PROXY_FILE_LONG, true, "Proxy list file")
        option.args = 1
        option.setOptionalArg(false)
        option.argName = "proxy options "
        options.addOption(option)

        option = Option(OPTION_WIDTH_SHORT, OPTION_WIDTH_LONG, true, "Width")
        option.args = 1
        option.setOptionalArg(true)
        option.argName = "Width "
        options.addOption(option)

        option = Option(OPTION_IMG_RELOAD_SHORT, OPTION_IMG_RELOAD_LONG, true, "Reload images")
        option.args = 0
        option.setOptionalArg(true)
        option.argName = "Reload images "
        options.addOption(option)

        option = Option(OPTION_SECURE_MODE_SHORT, OPTION_SECURE_MODE_LONG, true, "Secure mode")
        option.args = 0
        option.setOptionalArg(true)
        option.argName = "Secure mode "
        options.addOption(option)

        option = Option(OPTION_DEBUG_MODE_SHORT, OPTION_DEBUG_MODE_LONG, true, "Debug mode")
        option.args = 0
        option.setOptionalArg(true)
        option.argName = "Debug mode "
        options.addOption(option)

        option = Option(OPTION_PDF_MODE_SHORT, OPTION_PDF_MODE_LONG, true, "Pdf mode")
        option.args = 1
        option.setOptionalArg(true)
        option.argName = "Pdf mode "
        options.addOption(option)

        option = Option(OPTION_SCAN_MODE_SHORT, OPTION_SCAN_MODE_LONG, true, "Scan mode")
        option.args = 0
        option.setOptionalArg(true)
        option.argName = "Scan mode "
        options.addOption(option)

        option = Option(OPTION_CTX_MODE_SHORT, OPTION_CTX_MODE_LONG, true, "CTX mode")
        option.args = 2
        option.setOptionalArg(true)
        option.argName = "CTX mode "
        options.addOption(option)

        option = Option(OPTION_AUTH_MODE_SHORT, OPTION_AUTH_MODE_LONG, true, "authentication params")
        option.args = 2
        option.setOptionalArg(true)
        option.argName = "authentication params "
        options.addOption(option)

        commandLine = cmdLineParser.parse(options, commandLineArguments)
    }

    private fun getStringOptionValue(optionName: String): String {
        return if (commandLine.hasOption(optionName) && null != commandLine.getOptionValues(optionName) && 1 == commandLine.getOptionValues(optionName).size)
            commandLine.getOptionValues(optionName)[0]
        else
            ""
    }

    private fun getStringOptionValues(optionName: String): Array<String>? {
        return if (commandLine.hasOption(optionName) && null != commandLine.getOptionValues(optionName))
            commandLine.getOptionValues(optionName)
        else
            null
    }

    private fun getIntOptionValue(optionName: String): Int {
        return if (commandLine.hasOption(optionName) && 1 == commandLine.getOptionValues(optionName).size) Integer.parseInt(commandLine.getOptionValues(optionName)[0]) else 0
    }

    private fun getBoolOptionValue(optionName: String): Boolean {
        return commandLine.hasOption(optionName)
    }

    override val reloadImages: Boolean
        get() = getBoolOptionValue(OPTION_IMG_RELOAD_SHORT)

    override val secureMode: Boolean
        get() = getBoolOptionValue(OPTION_SECURE_MODE_SHORT)

    override val debugEnabled: Boolean
        get() = getBoolOptionValue(OPTION_DEBUG_MODE_SHORT)

    override val scanEnabled: Boolean
        get() = getBoolOptionValue(OPTION_SCAN_MODE_SHORT)

    override fun pdfOptions(): String {
        return getStringOptionValue(OPTION_PDF_MODE_SHORT)
    }

    override fun ctxOptions(): CtxOptions {
        val ctxOpts = getStringOptionValues(OPTION_CTX_MODE_SHORT)
        return if (ctxOpts == null || ctxOpts.size != 2) CtxOptions.DEFAULT_CTX_OPTIONS else CtxOptions(ctxOpts[0], ctxOpts[1])
    }

    override fun authOptions(): AuthOptions? {
        val authOpts = getStringOptionValues(OPTION_AUTH_MODE_SHORT)
        return if (authOpts == null || authOpts.size != 2) null else AuthOptions(authOpts[0], authOpts[1])
    }

    companion object {

        private const val OPTION_BOOKID_SHORT = "i"
        private const val OPTION_BOOKID_LONG = "bookId"
        private const val OPTION_OUTDIR_SHORT = "o"
        private const val OPTION_OUTDIR_LONG = "out"
        private const val OPTION_PROXY_FILE_SHORT = "p"
        private const val OPTION_PROXY_FILE_LONG = "proxy"
        private const val OPTION_WIDTH_SHORT = "w"
        private const val OPTION_WIDTH_LONG = "width"
        private const val OPTION_IMG_RELOAD_SHORT = "r"
        private const val OPTION_IMG_RELOAD_LONG = "reloadImages"
        private const val OPTION_SECURE_MODE_SHORT = "s"
        private const val OPTION_SECURE_MODE_LONG = "secure"
        private const val OPTION_PDF_MODE_SHORT = "x"
        private const val OPTION_PDF_MODE_LONG = "pdf"
        private const val OPTION_CTX_MODE_SHORT = "c"
        private const val OPTION_CTX_MODE_LONG = "ctx"
        private const val OPTION_AUTH_MODE_SHORT = "a"
        private const val OPTION_AUTH_MODE_LONG = "auth"
        private const val OPTION_DEBUG_MODE_SHORT = "d"
        private const val OPTION_DEBUG_MODE_LONG = "debug"
        private const val OPTION_SCAN_MODE_SHORT = "S"
        private const val OPTION_SCAN_MODE_LONG = "Scan"
    }
}
